llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -16624,15 +16624,55 @@ static SDValue reduceANDOfAtomicLoad(SDNode *N,
   return SDValue(N, 0);
 }
 
+// Sometimes a mask is applied after a shift. If that shift was fed by a
+// load, there is sometimes the opportunity to narrow the load, which is
+// hidden by the intermediate shift. Detect that case and commute the
+// shift/and in order to enable load narrowing.
+static SDValue combineNarrowableShiftedLoad(SDNode *N, SelectionDAG &DAG) {
+  // (and (shl (load ...), ShiftAmt), Mask)
+  using namespace SDPatternMatch;
+  SDValue LoadNode, ShiftNode;
+  APInt MaskVal, ShiftVal;
+  // (and (shl (load ...), ShiftAmt), Mask)
+  if (!sd_match(N,
+                m_And(m_OneUse(m_Shl(
+                          m_AllOf(m_Opc(ISD::LOAD), m_Value(LoadNode)),
+                          m_AllOf(m_ConstInt(ShiftVal), m_Value(ShiftNode)))),
```

Comment:
ShiftNode seems misnamed here - I think this is something like ShiftValNode?

And honestly, reuse of the existing node seems like slight overkill.  Using getShiftAmtConstant is more idiomatic.